src/Data/Multimap/Generic.hs

```diff
@@ -68,6 +73,25 @@ instance Foldable c => Foldable (Multimap c k) where
 instance Traversable c => Traversable (Multimap c k) where
   sequenceA (Multimap m) = Multimap <$> sequenceA (fmap sequenceA m)
 
+instance (Monoid k, Ord k, Collection c, Foldable c, Alternative c) => Applicative (Multimap c k) where
+  pure = singleton mempty
+  (<*>) = ap
+
+instance (Monoid k, Ord k, Collection c, Foldable c, Alternative c) => Alternative (Multimap c k) where
+  empty = unalt mempty
+  m1 <|> m2 = unalt (alt m1 <> alt m2)
+
+instance (Monoid k, Ord k, Collection c, Foldable c, Alternative c) => Monad (Multimap c k) where
+  return = pure
+  Multimap m >>= f = unalt $ joinMultimap (foldMap (alt . f) <$> m)
+    where joinMultimap = Map.foldMapWithKey (mapGroupKeysWith (<>) . (<>))
```

Comment:
These seem like unusual instances for a container type. Can you give practical use-cases for them?

Reply:
Nothing concrete at the moment, but the instances' behaviour would be a combination of `Writer` and `[]` i.e. what you get is a non-determinism with logging. All alternatives with the same log output get grouped together, because the log output is used as the map key. You'd pick the key type to represent an outcome of the computation that you care about. If the key type is `Any` for example, you'll end up with two lists of computations: one for computations whose any step produced `Any True` and one for all others.